src/index/sparse/sparse_inverted_index.h

```diff
@@ -0,0 +1,457 @@
+        std::fill(labels, labels + k, -1);
+        if (query.size() == 0) {
+            return;
+        }
+
+        // if no data was dropped during both build and search, no refinement is needed.
+        if (!drop_during_build_ && drop_ratio_search == 0) {
+            refine_factor = 1;
+        }
+
+        std::vector<T> values(query.size());
+        for (size_t i = 0; i < query.size(); ++i) {
+            values[i] = std::abs(query[i].val);
+        }
+        auto pos = values.begin() + static_cast<size_t>(drop_ratio_search * values.size());
+        std::nth_element(values.begin(), pos, values.end());
+        auto q_threshold = *pos;
+
+        MaxMinHeap<T> heap(k * refine_factor);
+        if (!use_wand_) {
+            search_brute_force(query, q_threshold, heap, bitset);
+        } else {
+            search_wand(query, q_threshold, heap, bitset);
+        }
+
+        if (refine_factor == 1) {
+            collect_result(heap, distances, labels);
+        } else {
+            refine_and_collect(query, heap, k, distances, labels);
+        }
+    }
+
+    void
+    GetVectorById(const label_t id, SparseRow<T>& output) const {
+        output.copy_from(raw_data_[id]);
+    }
+
+    [[nodiscard]] size_t
+    size() const {
+        std::shared_lock<std::shared_mutex> lock(mu_);
+        size_t res = sizeof(*this);
+        res += sizeof(SparseRow<T>) * n_rows_internal();
+        for (auto& row : raw_data_) {
+            res += row.byte_size();
+        }
+
+        res += sizeof(std::vector<IdVal<T>>) * inverted_lut_.capacity();
+        for (auto& lut : inverted_lut_) {
+            res += sizeof(IdVal<T>) * lut.capacity();
+        }
+        if (use_wand_) {
+            res += sizeof(T) * max_in_dim_.capacity();
+        }
+        return res;
+    }
+
+    [[nodiscard]] size_t
+    n_rows() const {
+        std::shared_lock<std::shared_mutex> lock(mu_);
+        return n_rows_internal();
+    }
+
+    [[nodiscard]] size_t
+    n_cols() const {
+        std::shared_lock<std::shared_mutex> lock(mu_);
+        return n_cols_internal();
+    }
+
+ private:
+    size_t
+    n_rows_internal() const {
+        return raw_data_.size();
+    }
+
+    size_t
+    n_cols_internal() const {
+        return inverted_lut_.size();
+    }
+
+    // find the top-k candidates using brute force search, k as specified by the capacity of the heap.
+    // any value in q_vec that is smaller than q_threshold and any value with dimension >= n_cols() will be ignored.
+    void
+    search_brute_force(const SparseRow<T>& q_vec, T q_threshold, MaxMinHeap<T>& heap, const BitsetView& bitset) const {
+        std::vector<float> scores(n_rows_internal(), 0.0f);
+        for (size_t idx = 0; idx < q_vec.size(); ++idx) {
+            auto [i, v] = q_vec[idx];
+            if (v < q_threshold || i >= n_cols_internal()) {
+                continue;
+            }
+            for (size_t j = 0; j < inverted_lut_[i].size(); j++) {
```

Comment:
this loop can be improved via SIMD. Maybe, later

Reply:
great suggestion! I'll leave that as is for now but I left a TODO.